### DISCUSSION TITLE
How to manage MicroStream's memory?

### DISCUSSION THREAD

**Discussion by nhthong2001** [microstream-one/microstream#508] [Q&A] [answered]
Hi all,
I am implementing a cache using MicroStream and Lazy Loading mechanism. 
I want to ask is there a way to know how much memory MicroStream is using on Ram while running and how to configure memory usage for MicroStream only? Is there any best practice for using MicroStream for caching? 

Thanks very much!!!

**Comment by hg-ms** [comment, marked as answer]
Hi,
First of all we have a JCache ([JSR-107](https://www.jcp.org/en/jsr/detail?id=107)) implementation, which is optionally backed by a MicroStream Storage (https://docs.microstream.one/manual/cache/index.html). Maybe this already meets your requirements.

An empty Microstream storage instance requires less then 10 MB memory. The more important part are the management data required for each persisted object instance.
To estimate the memory usage the most important part is the additional data required for each persisted object. You can calculate round about 188 Bytes per stored Object. 
LazyReferences require additional 104 Bytes even if unloaded. Using them with very small objects may counterproductive.
Microstream also uses an internal off-heap cache that can become quite large. If the default configuration of that cache is not suitable have look [here](https://docs.microstream.one/manual/storage/configuration/housekeeping.html#_cache).


Best regards

**Reply by nhthong2001** [reply]
Thanks for your reply.

Currently I have a server with a cache limit of 1GB. Now I want to configure MicroStream to only use within 1GB eg 512MB without OOM.

And "To estimate the memory usage the most important part is the additional data required for each persisted object. You can calculate round about 188 Bytes per stored Object."  --> Why 188 Bytes per stored Object? I'm not sure about this

**Reply by hg-ms** [reply]
There is no option to configure a maximum allowed memory usage for Microstream. If you are low on memory you may try to limit the number of persisted items and configure the [housekeeping](https://docs.microstream.one/manual/storage/housekeeping.html) more aggressive.
You may also [delete items](https://docs.microstream.one/manual/storage/deleting-data.html) from the storage. This is done be removing all references to an item. If an item is no more reachable in the persisted object-graph it will be deleted from Microstream whenever the “housekeeping” decides….

**Reply by nhthong2001** [reply]
And "To estimate the memory usage the most important part is the additional data required for each persisted object. You can calculate round about 188 Bytes per stored Object." --> Why 188 Bytes per stored Object? I'm not sure about this

**Reply by hg-ms** [reply]
Those 188 Bytes are (an estimation of) the overhead required by an active Microstream-storage to manage a persisted object.
